File: apps/kpi/utils/formatters.py
from decimal import Decimal
from typing import Union, Optional, Dict
from django.utils.html import format_html
from .decimal_utils import DecimalUtils

class FormatUtils:
    @classmethod
    def format_score(cls, score: Union[int, float, str, Decimal, None], 
                     decimal_places: int = 0, include_percent: bool = True) -> str:
        if score is None:
            return "N/A"
        score_dec = DecimalUtils.safe_decimal(score)
        formatted = DecimalUtils.round_decimal(score_dec, decimal_places)
        if include_percent:
            return f"{formatted}%"
        return str(formatted)
    
    @classmethod
    def format_currency(cls, value: Union[int, float, str, Decimal, None],
                        currency: str = "KES", decimal_places: int = 2) -> str:
        if value is None:
            return "N/A"
        value_dec = DecimalUtils.safe_decimal(value)
        formatted = DecimalUtils.round_decimal(value_dec, decimal_places)
        formatted_str = f"{formatted:,.{decimal_places}f}"
        return f"{currency} {formatted_str}"
    
    @classmethod
    def format_percentage(cls, value: Union[int, float, str, Decimal, None],
                          decimal_places: int = 1) -> str:
        if value is None:
            return "N/A"
        value_dec = DecimalUtils.safe_decimal(value)
        formatted = DecimalUtils.round_decimal(value_dec, decimal_places)
        return f"{formatted}%"
    
    @classmethod
    def format_traffic_light(cls, status: str, with_emoji: bool = True) -> str:
        config = {
            'GREEN': {'emoji': '🟢', 'label': 'On Track', 'color': '#22c55e'},
            'YELLOW': {'emoji': '🟡', 'label': 'At Risk', 'color': '#eab308'},
            'RED': {'emoji': '🔴', 'label': 'Off Track', 'color': '#ef4444'},
        }
        status_config = config.get(status, {'emoji': '⚪', 'label': 'Unknown', 'color': '#6b7280'})
        if with_emoji:
            return f"{status_config['emoji']} {status_config['label']}"
        return status_config['label']
    @classmethod
    def format_metric_value(cls, value: Union[int, float, str, Decimal, None],
                            unit: str = "", decimal_places: int = 0,
                            compact: bool = False) -> str:
        if value is None:
            return "N/A"
        value_dec = DecimalUtils.safe_decimal(value)
        if compact:
            if value_dec >= 1_000_000_000:
                formatted = f"{value_dec / 1_000_000_000:.1f}B"
            elif value_dec >= 1_000_000:
                formatted = f"{value_dec / 1_000_000:.1f}M"
            elif value_dec >= 1_000:
                formatted = f"{value_dec / 1_000:.1f}K"
            else:
                formatted = DecimalUtils.round_decimal(value_dec, decimal_places)
                formatted = f"{formatted:,}"
        else:
            formatted = DecimalUtils.round_decimal(value_dec, decimal_places)
            formatted = f"{formatted:,}"
        if unit:
            return f"{formatted} {unit}"
        return formatted
    
    @classmethod
    def format_period(cls, year: int, month: int, format_type: str = 'YYYY-MM') -> str:
        """Format period string"""
        if format_type == 'YYYY-MM':
            return f"{year}-{month:02d}"
        elif format_type == 'MM/YYYY':
            return f"{month:02d}/{year}"
        elif format_type == 'Month YYYY':
            from .date_utils import get_month_name
            return f"{get_month_name(month)} {year}"
        elif format_type == 'MMM YYYY':
            from .date_utils import get_month_name
            return f"{get_month_name(month, abbreviated=True)} {year}"
        return f"{year}-{month:02d}"
    
    @classmethod
    def format_trend(cls, trend_data: Dict) -> str:
        direction = trend_data.get('direction', 'STABLE')
        confidence = trend_data.get('confidence', 0)
        config = {
            'IMPROVING': {'arrow': '↑', 'color': '#22c55e'},
            'DECLINING': {'arrow': '↓', 'color': '#ef4444'},
            'STABLE': {'arrow': '→', 'color': '#6b7280'},
            'VOLATILE': {'arrow': '↗↘', 'color': '#eab308'},
        }
        cfg = config.get(direction, {'arrow': '?', 'color': '#6b7280'})
        return f"<span style='color: {cfg['color']}'>{cfg['arrow']}</span>"
    
    @classmethod
    def format_duration(cls, seconds: float) -> str:
        if seconds < 60:
            return f"{seconds:.0f}s"
        elif seconds < 3600:
            minutes = seconds / 60
            return f"{minutes:.1f}m"
        else:
            hours = seconds / 3600
            return f"{hours:.1f}h"
    
    @classmethod
    def format_number(cls, value: Union[int, float, str, Decimal, None],
                      decimal_places: int = 0) -> str:
        if value is None:
            return "N/A"
        value_dec = DecimalUtils.safe_decimal(value)
        formatted = DecimalUtils.round_decimal(value_dec, decimal_places)
        return f"{formatted:,.{decimal_places}f}"

def format_score(score: Union[int, float, str, Decimal, None], 
                 decimal_places: int = 0, include_percent: bool = True) -> str:
    return FormatUtils.format_score(score, decimal_places, include_percent)

def format_currency(value: Union[int, float, str, Decimal, None],
                    currency: str = "KES", decimal_places: int = 2) -> str:
    return FormatUtils.format_currency(value, currency, decimal_places)

def format_percentage(value: Union[int, float, str, Decimal, None],
                      decimal_places: int = 1) -> str:
    return FormatUtils.format_percentage(value, decimal_places)

def format_traffic_light(status: str, with_emoji: bool = True) -> str:
    return FormatUtils.format_traffic_light(status, with_emoji)

def format_metric_value(value: Union[int, float, str, Decimal, None],
                        unit: str = "", decimal_places: int = 0,
                        compact: bool = False) -> str:
    return FormatUtils.format_metric_value(value, unit, decimal_places, compact)

def format_period(year: int, month: int, format_type: str = 'YYYY-MM') -> str:
    return FormatUtils.format_period(year, month, format_type)

File: frontend/src/utils/kpi/formatters.js
import { TRAFFIC_LIGHT, PRECISION, DATE_FORMATS } from './constants';

export const formatScore = (score, decimals = PRECISION.SCORE) => {
    if (score === null || score === undefined || isNaN(score)) return 'N/A';
    return `${score.toFixed(decimals)}%`;
};
export const formatCurrency = (value, currency = 'KES', decimals = PRECISION.CURRENCY) => {
    if (value === null || value === undefined || isNaN(value)) return 'N/A';
    return new Intl.NumberFormat('en-KE', {
        style: 'currency',
        currency: currency,
        minimumFractionDigits: decimals,
        maximumFractionDigits: decimals
    }).format(value);
};
export const formatNumber = (value, decimals = 0) => {
    if (value === null || value === undefined || isNaN(value)) return 'N/A';
    return new Intl.NumberFormat('en-US', {
        minimumFractionDigits: decimals,
        maximumFractionDigits: decimals
    }).format(value);
};
export const formatPercentage = (value, decimals = PRECISION.PERCENTAGE) => {
    if (value === null || value === undefined || isNaN(value)) return 'N/A';
    return `${value.toFixed(decimals)}%`;
};
export const formatTrafficLight = (status, withEmoji = true) => {
    const config = TRAFFIC_LIGHT[status];
    if (!config) return 'Unknown';
    if (withEmoji) return `${config.emoji} ${config.label}`;
    return config.label;
};
export const getTrafficLightColor = (status) => {
    return TRAFFIC_LIGHT[status]?.color || '#6b7280';
};
export const getTrafficLightFromScore = (score) => {
    if (score >= TRAFFIC_LIGHT.GREEN.threshold) return TRAFFIC_LIGHT.GREEN;
    if (score >= TRAFFIC_LIGHT.YELLOW.threshold) return TRAFFIC_LIGHT.YELLOW;
    return TRAFFIC_LIGHT.RED;
};

export const formatValidationStatus = (status) => {
    return VALIDATION_STATUS[status] || { label: status, color: '#6b7280', icon: '📝' };
};
export const formatKPIType = (type) => {
    return KPI_TYPES[type]?.label || type;
};

export const formatCalculationLogic = (logic) => {
    return CALCULATION_LOGIC[logic]?.label || logic;
};

export const formatMeasureType = (type) => {
    return MEASURE_TYPES[type]?.label || type;
};
export const formatMetricValue = (value, unit = '', type = null, decimals = PRECISION.DEFAULT) => {
    if (value === null || value === undefined || isNaN(value)) return 'N/A';
    
    let formatted = value.toFixed(decimals);
    formatted = formatted.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
    
    if (type === 'PERCENTAGE') return `${formatted}%`;
    if (type === 'FINANCIAL') return formatCurrency(value);
    if (unit) return `${formatted} ${unit}`;
    return formatted;
};
export const formatDuration = (seconds) => {
    if (!seconds || isNaN(seconds)) return '0s';
    
    if (seconds < 60) {
        return `${Math.round(seconds)}s`;
    } else if (seconds < 3600) {
        const minutes = seconds / 60;
        return `${minutes.toFixed(1)}m`;
    } else {
        const hours = seconds / 3600;
        return `${hours.toFixed(1)}h`;
    }
};
export const formatFileSize = (bytes) => {
    if (bytes === 0) return '0 Bytes';
    const k = 1024;
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
};
export const formatRelativeTime = (date) => {
    if (!date) return 'N/A';
    const now = new Date();
    const past = new Date(date);
    const diffMs = now - past;
    const diffSec = Math.floor(diffMs / 1000);
    const diffMin = Math.floor(diffSec / 60);
    const diffHour = Math.floor(diffMin / 60);
    const diffDay = Math.floor(diffHour / 24);
    const diffMonth = Math.floor(diffDay / 30);
    const diffYear = Math.floor(diffDay / 365);
    if (diffSec < 60) return 'just now';
    if (diffMin < 60) return `${diffMin} minute${diffMin > 1 ? 's' : ''} ago`;
    if (diffHour < 24) return `${diffHour} hour${diffHour > 1 ? 's' : ''} ago`;
    if (diffDay < 30) return `${diffDay} day${diffDay > 1 ? 's' : ''} ago`;
    if (diffMonth < 12) return `${diffMonth} month${diffMonth > 1 ? 's' : ''} ago`;
    return `${diffYear} year${diffYear > 1 ? 's' : ''} ago`;
};
export const truncateText = (text, length = 100) => {
    if (!text) return '';
    if (text.length <= length) return text;
    return text.substring(0, length) + '...';
};
export const formatTrend = (direction) => {
    const trends = {
        IMPROVING: { icon: '📈', label: 'Improving', color: '#22c55e' },
        DECLINING: { icon: '📉', label: 'Declining', color: '#ef4444' },
        STABLE: { icon: '➡️', label: 'Stable', color: '#6b7280' },
        VOLATILE: { icon: '↗↘', label: 'Volatile', color: '#eab308' }
    };
    return trends[direction] || { icon: '📊', label: 'Unknown', color: '#6b7280' };
};